File: src/app/model/CurrentTimeMetric.ts
import { Metric, MetricType } from "./Metric";
import { MetricService } from "../services/metric.service";
import { UnitToString } from "../util/unit-to-string";

export class CurrentTimeMetric extends Metric {
    private intervalId: any;

    constructor(metricService: MetricService) {
        super(MetricType.CURRENT_TIME, 'Current Time', 's', metricService, 's', 0);
    }

    startLogging(): void {
        if (this.intervalId) return;
        this.intervalId = setInterval(() => {
            const now = new Date();
            this.addValue(now.getTime() / 1000, now);
        }, 1000); // Log every second
    }

    stopLogging(): void {
        if (this.intervalId) {
            clearInterval(this.intervalId);
            this.intervalId = null;
        }
    }

    protected override displayValue(value: number | null, includeUnit: boolean = true): string {
        if (value === null) return '---';
        const date = new Date(value * 1000); // Convert seconds to milliseconds
        const hours = date.getHours().toString().padStart(2, '0');
        const minutes = date.getMinutes().toString().padStart(2, '0');
        const formattedTime = `${hours}:${minutes}`;
        return formattedTime;
    }

}

File: android/app/src/main/java/com/openbikecomputer/model/CurrentTimeMetric.java
package com.openbikecomputer.model;

import com.openbikecomputer.service.IMetricService;

import java.util.Calendar;
import java.util.Locale;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class CurrentTimeMetric extends Metric {
    private ScheduledExecutorService executorService;
    private boolean started = false;

    public CurrentTimeMetric(IMetricService metricService) {
        super(MetricType.CURRENT_TIME, "Current Time", "s", metricService, "s", 0);
    }

    @Override
    public void startLogging() {
        if (started) {
            return;
        }
        started = true;
        executorService = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread thread = new Thread(r, "CurrentTimeMetric");
            thread.setDaemon(true);
            return thread;
        });
        executorService.scheduleAtFixedRate(() -> {
            long now = System.currentTimeMillis();
            addValue(now / 1000.0, now);
        }, 1000, 1000, TimeUnit.MILLISECONDS);
    }

    @Override
    public void stopLogging() {
        if (executorService != null) {
            executorService.shutdownNow();
            executorService = null;
        }
        started = false;
    }

    @Override
    protected String displayValue(Double value, boolean includeUnit) {
        if (value == null) {
            return "---";
        }
        Calendar date = Calendar.getInstance();
        date.setTimeInMillis((long) (value * 1000));
        return String.format(Locale.US, "%02d:%02d",
                date.get(Calendar.HOUR_OF_DAY),
                date.get(Calendar.MINUTE));
    }
}
